File: Leetcode_All_Solved/547.cpp
class Solution {
    int n;
    int set_find(vector<int>& set, int a){
        if(set[a] == (-1))return a;
        return (set[a] = set_find(set, set[a]));
    }
    void set_join(vector<int>& set, int a, int b){
        a = set_find(set, a);
        b = set_find(set, b);
        if(a != b) set[a] = b;
    }
    inline int n_set(const vector<int>& set){
        int ans = 0;
        for(int root:set)if(root == (-1))ans++;
        return ans;
    }
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        n = isConnected.size();
        vector<int> set(n, (-1));
        
        for(int i=0;i<n;i++)for(int j=i+1;j<n;j++)if(isConnected[i][j])
            set_join(set,i,j);
        return n_set(set);
    }
};

File: Leetcode_All_Solved/547.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        visited = [False for i in range(n)]

        al = [[] for i in range(n)]
        for i in range(n):
            for j in range(i+1, n):
                if(isConnected[i][j]):
                    al[i].append(j)
                    al[j].append(i)
        
        def dfs(al: List[List[int]], u: int) -> None:
            if(visited[u] == True):
                return
            visited[u] = True

            for v in al[u]:
                dfs(al, v)
            return

        component = 0
        for i in range(n):
            if(visited[i] == False):
                dfs(al, i)
                component += 1
        return component
